File: SOC/src/cc/game/soc/core/DistancesLand.java
package cc.game.soc.core;

import java.util.ArrayList;
import java.util.List;

public class DistancesLand implements IDistances {

	public DistancesLand() {
		this(null, null);
	}
	
	private final byte [][] dist;
	private final byte [][] next;
	
	public DistancesLand(byte [][] dist, byte [][] next) {
		this.dist = dist;
		this.next = next;
	}
	
	public List<Integer> getShortestPath(int fromVertex, int toVertex) {
		List<Integer> path = new ArrayList<Integer>();
		if (dist[fromVertex][toVertex] != DISTANCE_INFINITY)
		{
			path.add(fromVertex);
			while (fromVertex != toVertex) {
				fromVertex = next[fromVertex][toVertex];
				path.add(fromVertex);
			}
		}		
		return path;
	}
	
	public int getDist(int from, int to) {
		return dist[from][to];
	}

}

File: SOC/src/cc/game/soc/core/DistancesLand.kt
package cc.game.soc.core

import java.util.*

class DistancesLand @JvmOverloads constructor(private val dist: Array<ByteArray> = arrayOf(), private val next: Array<ByteArray> = arrayOf()) : IDistances {
	override fun getShortestPath(fromVertex: Int, toVertex: Int): List<Int> {
		var fromVertex = fromVertex
		val path: MutableList<Int> = ArrayList()
		if (dist[fromVertex][toVertex] != IDistances.DISTANCE_INFINITY) {
			path.add(fromVertex)
			while (fromVertex != toVertex) {
				fromVertex = next!![fromVertex][toVertex].toInt()
				path.add(fromVertex)
			}
		}
		return path
	}

	override fun getDist(from: Int, to: Int): Int {
		return dist[from][to].toInt()
	}
}